clients/src/main/java/org/apache/kafka/common/utils/Bytes.java

```diff
@@ -0,0 +1,415 @@
+     * @param length1 How much to compare from the left buffer
+     * @param length2 How much to compare from the right buffer
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static int compareTo(final byte[] buffer1, final int offset1, final int length1,
+                                final byte[] buffer2, final int offset2, final int length2) {
+
+        return ((ByteArrayComparator) BYTES_LEXICO_COMPARATOR).compare(buffer1, offset1, length1, buffer2, offset2, length2);
+    }
+
+    /**
+     * Check if two byte arrays are lexicographically equal to each other.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static boolean equals(final byte[] buffer1, final byte[] buffer2) {
+        // short circuit case, other shortcuts are in the other equals function
+        if (buffer1 == buffer2) return true;
+
+        return equals(buffer1, 0, buffer1.length, buffer2, 0, buffer2.length);
+    }
+
+    /**
+     * Check if two byte arrays are lexicographically equal to each other.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @param offset1 Where to start comparing in the left buffer
+     * @param offset2 Where to start comparing in the right buffer
+     * @param length1 How much to compare from the left buffer
+     * @param length2 How much to compare from the right buffer
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static boolean equals(final byte[] buffer1, int offset1, int length1,
+                                 final byte[] buffer2, int offset2, int length2) {
+        // short circuit case
+        if (buffer1 == null || buffer2 == null)
+            return false;
+        if (buffer1 == buffer2 &&
+                offset1 == offset2 &&
+                length1 == length2)
+            return true;
+        if (length1 != length2)
+            return false;
+        if (length1 == 0)
+            return true;
+
+        // since we're often comparing adjacent sorted data,
+        // it's usual to have equal arrays except for the very last byte
+        // so check that first
+        if (buffer1[offset1 + length1 - 1] != buffer2[offset2 + length2 - 1]) return false;
+
+        return ((ByteArrayComparator) BYTES_LEXICO_COMPARATOR).compare(buffer1, offset1, length1, buffer2, offset2, length2) == 0;
+    }
+
+    /**
+     * Returns a new byte array, copied from the given {@code buf},
+     * from the position (inclusive) to the limit (exclusive).
+     * The position and the other index parameters are not changed.
+     *
+     * @param buf a byte buffer
+     * @return the byte array
+     * @see #toBytes(ByteBuffer)
+     */
+    public static byte[] getBytes(ByteBuffer buf) {
+        byte [] result = new byte[buf.remaining()];
+        buf.get(result);
```

Comment:
It may not a problem given the current usage, but thought I'd point out that the comment on this particular method conflicts with its implementation ("The position and the other index parameters are not changed").